src/clients/c++/examples/simple_cuda_shm_client.cc

```diff
@@ -0,0 +1,398 @@
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to register shared memory output region: "
+              << err << std::endl;
+    exit(1);
+  }
+
+  // Set the context options to do batch-size 1 requests. Also request that
+  // all output tensors be returned using shared memory.
+  std::unique_ptr<nic::InferContext::Options> options;
+  FAIL_IF_ERR(
+      nic::InferContext::Options::Create(&options),
+      "unable to create inference options");
+
+  options->SetBatchSize(1);
+  options->AddSharedMemoryResult(output0, "output_data", 0, output_byte_size);
+  options->AddSharedMemoryResult(
+      output1, "output_data", output_byte_size, output_byte_size);
+
+  FAIL_IF_ERR(
+      infer_ctx->SetRunOptions(*options), "unable to set inference options");
+
+  // Create Output0 and Output1 in CUDA Shared Memory. Initialize Input0 to
+  // unique integers and Input1 to all ones.
+  int input_data[32];
+  for (size_t i = 0; i < 16; ++i) {
+    input_data[i] = i;
+    input_data[16 + i] = 1;
+  }
+
+  // copy INPUT0 and INPUT1 data in GPU shared memory
+  int *input0_d_ptr, *input1_d_ptr;
+  cudaMalloc((void**)&input0_d_ptr, input_byte_size * 2);
+  cudaMemcpy(
+      (void*)input0_d_ptr, (void*)input_data, input_byte_size * 2,
+      cudaMemcpyHostToDevice);
+  input1_d_ptr = (int*)input0_d_ptr + 16;
+
+  cudaIpcMemHandle_t input_cuda_handle;
+  CreateCUDAIPCHandle(&input_cuda_handle, (void*)input0_d_ptr);
+
+  // Register Input shared memory with TRTIS
+  err = shared_memory_ctx->RegisterCudaSharedMemory(
+      "input_data", input_cuda_handle, input_byte_size * 2, 0);
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to register shared memory input region: " << err
+              << std::endl;
+    exit(1);
+  }
+
+  // Set the shared memory region for Inputs
+  err = input0->SetSharedMemory("input_data", 0, input_byte_size);
+  if (!err.IsOk()) {
+    std::cerr << "failed setting shared memory input: " << err << std::endl;
+    exit(1);
+  }
+  err = input1->SetSharedMemory("input_data", input_byte_size, input_byte_size);
+  if (!err.IsOk()) {
+    std::cerr << "failed setting shared memory input: " << err << std::endl;
+    exit(1);
+  }
+
+  // Send inference request to the inference server.
+  std::map<std::string, std::unique_ptr<nic::InferContext::Result>> results;
+  FAIL_IF_ERR(infer_ctx->Run(&results), "unable to run model");
+
+  // We expect there to be 2 results. Walk over all 16 result elements
+  // and print the sum and difference calculated by the model.
+  if (results.size() != 2) {
+    std::cerr << "error: expected 2 results, got " << results.size()
+              << std::endl;
+  }
+
+  // Copy input and output data back to the CPU
+  int input0_data[16], input1_data[16];
+  int output0_data[16], output1_data[16];
+  cudaMemcpy(
+      input0_data, input0_d_ptr, input_byte_size, cudaMemcpyDeviceToHost);
+  cudaMemcpy(
+      input1_data, input1_d_ptr, input_byte_size, cudaMemcpyDeviceToHost);
```

Comment:
We already have the input data in system memory (`input_data`)

Reply:
Removed